File: instrumentation/wireframe/src/main/java/io/opentelemetry/android/mobile/instrumentation/WireframeConfig.kt
// Copyright 2025 Barry Solomon
// SPDX-License-Identifier: Apache-2.0

package io.opentelemetry.android.mobile.instrumentation

/**
 * Configuration for [WireframeInstrumentation].
 *
 * Controls what is captured in the view-hierarchy wireframe and when captures are triggered.
 * The wireframe is a compact JSON tree (~1–5 KB per frame) describing the layout geometry
 * of every visible view, suitable for journey replay and structural analysis.
 *
 * @property enabled Whether wireframe capture is active.
 * @property captureOnScreenView Automatically capture a wireframe on every screen transition
 *   (pairs with [ScreenViewInstrumentation] events). This is the primary journey-building trigger.
 * @property captureOnTap Capture a wireframe after each tap event. **Off by default**:
 *   taps rarely change the view hierarchy, so per-tap captures produce duplicate JSON.
 *   Prefer associating taps with the most recent wireframe via the `mobile.wireframe.id`
 *   attribute (auto-attached by [TapInstrumentation] when this module is installed) and
 *   relying on [dedupeByContentHash] to suppress identical re-captures.
 * @property captureOnError Capture a wireframe when an uncaught exception occurs.
 * @property captureOnPolicyMatch Capture a wireframe whenever a buffered-export policy
 *   fires (crash-recovery, ui-freeze, http-error, etc.). The wireframe gets emitted into
 *   the same flush window, giving every server-side incident an attached "what was on
 *   screen" artifact. Default ON.
 * @property dedupeByContentHash When ON, the module hashes each captured wireframe and
 *   skips emit if the hash matches the previous capture — typical no-op taps and repeated
 *   `onResume` callbacks won't fill the buffer with duplicate JSON trees. Instead a
 *   `ui.wireframe.ref` log is emitted carrying the prior `mobile.wireframe.id`.
 * @property maxDepth Maximum view-tree traversal depth. Deeper trees are truncated with a
 *   `"truncated": true` marker. Keeps payload predictable for deeply nested layouts.
 * @property includeResourceIds Include Android resource IDs (e.g., `"btn_book"`) on views
 *   that have them. Useful for mapping wireframes to code, but may leak internal naming.
 * @property includeTextHints Include placeholder/hint text (e.g., "Enter name") but never
 *   actual user-entered text. Helps identify fields without exposing PII.
 * @property includeContentDescription Include accessibility `contentDescription` values.
 *   Helpful for understanding unlabelled icons/images.
 * @property includeClickableState Include `clickable` and `enabled` flags on each view.
 * @property maxCapturesPerMinute Rate limit to prevent excessive captures during rapid navigation.
 */
@Incubating
data class WireframeConfig(
    val enabled: Boolean = true,
    val captureOnScreenView: Boolean = true,
    val captureOnTap: Boolean = false,
    val captureOnError: Boolean = true,
    val captureOnPolicyMatch: Boolean = true,
    val dedupeByContentHash: Boolean = true,
    val maxDepth: Int = 30,
    val includeResourceIds: Boolean = true,
    val includeTextHints: Boolean = false,
    val includeContentDescription: Boolean = false,
    val includeClickableState: Boolean = true,
    val maxCapturesPerMinute: Int = 30
) {
    init {
        require(maxDepth in 1..100) { "maxDepth must be in 1..100" }
        require(maxCapturesPerMinute in 1..120) { "maxCapturesPerMinute must be in 1..120" }
    }
}

File: otel-ios-mobile/Sources/WireframeInstrumentation/WireframeConfig.swift
import Foundation
import OTelMobileCore

public struct WireframeConfig: Sendable {
    public var enabled: Bool
    public var maxCapturesPerMinute: Int
    public var maxDepth: Int
    public var captureOnScreenView: Bool
    public var captureOnTap: Bool
    public var captureOnError: Bool
    /// Capture a wireframe whenever a buffered-export policy fires
    /// (crash-recovery, ui-freeze, http-error, etc.). The wireframe gets
    /// emitted into the same flush window, giving every server-side incident
    /// an attached "what was on screen" artifact. Default `true`. Mirrors
    /// Android's `WireframeConfig.captureOnPolicyMatch`.
    public var captureOnPolicyMatch: Bool
    /// When `true` (default), the module hashes each captured wireframe JSON
    /// and emits a lightweight `ui.wireframe.ref` log carrying only the
    /// prior `mobile.wireframe.id` if the hash matches the previous capture.
    /// Saves the 1–5 KB payload on no-op screen-resume / tap captures.
    /// Mirrors Android's `WireframeConfig.dedupeByContentHash`.
    public var dedupeByContentHash: Bool
    public var includeAccessibilityIdentifiers: Bool
    public var includeTextHints: Bool
    public var includeContentDescription: Bool
    public var includeInteractionState: Bool
    /// Optional customer-owned consent gate. When non-`nil`, it is consulted
    /// **synchronously on the main thread immediately before each capture**
    /// (after the `enabled` and rate-limit checks, before the view-tree walk).
    /// If it returns `false`, the capture is skipped entirely — no walk, no
    /// log emitted. When `nil`, capture follows ``enabled`` alone. The consent
    /// gate is an *additional* runtime gate on top of ``enabled``, not a
    /// replacement. Keep the closure cheap and non-blocking; it runs in the
    /// capture hot path on the main thread.
    public var shouldCapture: CaptureConsentGate?

    public init(
        enabled: Bool = true,
        maxCapturesPerMinute: Int = 30,
        maxDepth: Int = 20,
        captureOnScreenView: Bool = true,
        captureOnTap: Bool = false,
        captureOnError: Bool = true,
        captureOnPolicyMatch: Bool = true,
        dedupeByContentHash: Bool = true,
        includeAccessibilityIdentifiers: Bool = true,
        includeTextHints: Bool = false,
        includeContentDescription: Bool = true,
        includeInteractionState: Bool = true,
        shouldCapture: CaptureConsentGate? = nil
    ) {
        self.enabled = enabled
        self.maxCapturesPerMinute = maxCapturesPerMinute
        self.maxDepth = maxDepth
        self.captureOnScreenView = captureOnScreenView
        self.captureOnTap = captureOnTap
        self.captureOnError = captureOnError
        self.captureOnPolicyMatch = captureOnPolicyMatch
        self.dedupeByContentHash = dedupeByContentHash
        self.includeAccessibilityIdentifiers = includeAccessibilityIdentifiers
        self.includeTextHints = includeTextHints
        self.includeContentDescription = includeContentDescription
        self.includeInteractionState = includeInteractionState
        self.shouldCapture = shouldCapture
    }

    /// Builder-style setter for the consent gate. Returns a copy with
    /// ``shouldCapture`` set.
    public func withConsentGate(_ gate: @escaping CaptureConsentGate) -> WireframeConfig {
        var copy = self
        copy.shouldCapture = gate
        return copy
    }
}
